### ISSUE TITLE
[HOLD for payment 2023-04-26] [$2000] New Room Dropdown does not properly reflect multiple spaces in the New Room Modal

### ISSUE THREAD

**Issue by alex-mechler** [Expensify/App#16329]
Coming from https://github.com/Expensify/App/pull/15646#discussion_r1132003674

## Action Performed:
1. Create a new workspace
2. Update the name to have multiple consecutive spaces 
3. Click the FAB
4. Select `New Room`
5. Observe the workspace name in the modal/dropdown doesn't show multiple spaces

## Expected Result:
The displayed name should show multiple spaces, same as elsewhere

## Actual Result:
The display name shows a single space
<img width="1673" alt="image" src="https://user-images.githubusercontent.com/22447860/226482946-1640b7c6-d739-42b8-b0a5-2e5ba01aced6.png">

## Workaround:
N/A, its simply a display issue

## Platforms:
Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [x] Android / Chrome
- [ ] iOS / native
- [x] iOS / Safari
- [x] MacOS / Chrome / Safari
- [x] MacOS / Desktop

**Version Number:** v1.2.84-1 DEV
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):** N/a
**Logs:** https://stackoverflow.com/c/expensify/questions/4856 N/a
**Notes/Photos/Videos:** Any additional supporting documentation
<img width="1673" alt="image" src="https://user-images.githubusercontent.com/22447860/226482946-1640b7c6-d739-42b8-b0a5-2e5ba01aced6.png">

**Expensify/Expensify Issue URL:** N/a
**Issue reported by:** @tienifr 
**Slack conversation:** N/a https://github.com/Expensify/App/pull/15646#discussion_r1132003674

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~018f5eeb68bbcf95bc</li>
        <li>Upwork Job ID: 1638322484300886016</li>
        <li>Last Price Increase: 2023-03-28</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Triggered auto assignment to @jliexpensify (`Bug`), see https://stackoverflow.com/c/expensify/questions/14418 for more details.

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by jliexpensify** [comment]
@alex-mechler I wanted to ask about this as I'm on v1.2.87-0. This is what I see when I input multiple spaces between words:

![image](https://user-images.githubusercontent.com/30815269/226490016-ab07800b-2cee-447e-b725-e707a7edba87.png)

I also don't have a DEV environment - can I actually test this?

**Comment by alex-mechler** [comment]
When https://github.com/Expensify/App/pull/15646 is actually deployed to staging, you should see multiple spaces in the workspace name. I created this issue because we know we needed the follow up after the pr was merged.

With the next app deploy you should be able to test this on staging

**Comment by jliexpensify** [comment]
Awesome, thanks for the context - will check it out once it's deployed!

**Comment by MelvinBot** [comment]
Current assignee @jliexpensify is eligible for the External assigner, not assigning anyone new.

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor-plus team member for initial proposal review - @0xmiroslav (`External`)

**Comment by MelvinBot** [comment]
Current assignee @alex-mechler is eligible for the External assigner, not assigning anyone new.

**Comment by dayyad** [comment]
## Proposal
### Problem Statement

In the current React Native application, when a user updates the workspace name with multiple consecutive spaces and opens the modal/dropdown for creating a new room, the workspace name is displayed with only a single space instead of the intended multiple consecutive spaces.
Root Cause

The root cause of this problem is that HTML automatically collapses multiple consecutive spaces into a single space when rendering text. As a result, the workspace name with multiple spaces is not being displayed as intended in the modal/dropdown.

### Proposed Solution

To solve this issue, we should wrap the workspace name text in a <span> element and apply the white-space: pre-wrap style. This style will preserve the white space in the text, including multiple consecutive spaces.

    Add a new style to the component in the file where the workspace name is displayed in the modal/dropdown

```
const styles = StyleSheet.create({
    // ...
    workspaceName: {
        whiteSpace: 'pre-wrap',
    },
});
```
    Wrap the workspace name text in a <span> element and apply the new style:

```
<Text style={styles.workspaceName}>
    <span style={styles.workspaceName}>{this.props.workspaceName}</span>
</Text>
```
This solution ensures that the workspace name is displayed with multiple consecutive spaces as intended in the modal/dropdown.
Alternative Solutions (Optional)

An alternative solution could involve using non-breaking spaces (&nbsp;) to replace multiple consecutive spaces. However, this approach requires additional processing of the input text and might introduce unnecessary complexity to the codebase.

**Comment by hellohublot** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

New Room Dropdown does not properly reflect multiple spaces in the New Room Modal

### What is the root cause of that problem?

select.option will filter consecutive spaces

### What changes do you think we should make in order to solve the problem?

https://github.com/Expensify/App/blob/ebc03dbb14d1689e35301278833a82716ae53034/src/components/Picker/Picker.js#L194-L195
``` diff
- // We add a text color to prevent white text on white background dropdown items on Windows
- items={_.map(this.props.items, item => ({...item, color: themeColors.pickerOptionsTextColor}))}
+ items={_.map(this.props.items, item => ({
+     ...item,
+     // We add a text color to prevent white text on white background dropdown items on Windows
+     color: themeColors. pickerOptionsTextColor,

+     // We replace normal space to unicode consecutive space to display consecutive space in browser, it's also support native parse
+     label: item.label.replace(/ /g, '\u00A0')
+ }))}
```

We replace it with `&nbsp;`, of course we need to use its unicode character `\u00A0`
If necessary, it can be written into `CONST.js` or `expensify-common`

I tried ['\n', '\t', '\r', ';', '<', '&'], but I have not found that other characters need to be replaced.
It works fine on the `mobile native` platform as it's a `unicode space`

### What alternative solutions did you explore? (Optional)
not yet

**Comment by koko57** [comment]
Can investigate this one :)

**Comment by MelvinBot** [comment]
📣 @koko57 You have been assigned to this job by @mountiny!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by jliexpensify** [comment]
Not overdue!

p.s. @tienifr - can you update your GH profile to include your full name and send me your Upworks profile so I can hire you for the reporting bonus? Thanks in advance!

**Comment by MelvinBot** [comment]
📣 @OlimpiaZurek You have been assigned to this job by @alex-mechler!
Please apply to this job in Upwork and leave a comment on the Github issue letting us know when we can expect a PR to be ready for review 🧑‍💻
Keep in mind: [Code of Conduct](https://github.com/Expensify/App/blob/main/CODE_OF_CONDUCT.md) | [Contributing](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md) 📖

**Comment by alex-mechler** [comment]
> @alex-mechler What class of solutions would you consider acceptable? For example, do "less hacky solutions" include switching to a custom HTML dropdown, or is that beyond the scope of this issue?
> Do we aim for displaying the spaces as-they-are in the native <option> HTML elements, which correspond to a browser-rendered native platform-dependent dropdown?

We would like to avoid platform specific code if possible. Switching to a custom drop down would not necessarily be outside the scope of this issue, but it sounds like a large undertaking. Ideally we would want something that displays the spaces as-is without modifying the text as we will miss cases where we have to modify the text. We also want to avoid platform specific code to avoid having inconsistencies across different platform

**Comment by alex-mechler** [comment]
> Is this list correct? Especially "Android / native" and "iOS / native"?

Just double checked this list, the two native platforms were errantly checked. All others were properly marked as being reproducible

As for your proposal, this issue is currently with @OlimpiaZurek, and does not have the `Help Wanted` label, so we aren't accepting proposals at this time. From https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job

`Do not propose solutions to jobs until the Help Wanted label has been applied. Any proposals submitted before these labels are added will not be reviewed.`

**Comment by tienifr** [comment]
@jliexpensify here's my Upwork profile https://www.upwork.com/freelancers/~01991fd5e5c11ef3ba, thanks!

**Comment by cubuspl42** [comment]
> > Is this list correct? Especially "Android / native" and "iOS / native"?
> 
> Just double checked this list, the two native platforms were errantly checked. All others were properly marked as being reproducible
> 
> As for your proposal, this issue is currently with @OlimpiaZurek, and does not have the `Help Wanted` label, so we aren't accepting proposals at this time. From https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#propose-a-solution-for-the-job
> 
> `Do not propose solutions to jobs until the Help Wanted label has been applied. Any proposals submitted before these labels are added will not be reviewed.`

I thought that that rule means that **once** that label is applied, this condition is satisfied, and removing it during Callstack assignment is another thing. That is what I understood from [this explanation](https://github.com/Expensify/App/issues/15648#issuecomment-1476374774) from another issue. Is everybody on the same page in this matter?

**Comment by alex-mechler** [comment]
> I thought that that rule means that once that label is applied, this condition is satisfied, and removing it during Callstack assignment is another thing (i.e is done not to attract new attention). That is what I understood from https://github.com/Expensify/App/issues/15648#issuecomment-1476374774 from another issue. Is everybody on the same page in this matter?

Ah, apologies here, this is the first time I've been the CME on a callstack issue, so I hadn't seen this discussion, I have talked to Vit directly and cleared up my confusion here. We can continue to look at proposals at this time.


@0xmiroslav do you have any thoughts on [this proposal](https://github.com/Expensify/App/issues/16329#issuecomment-1490525854)?

**Comment by jliexpensify** [comment]
@alex-mechler and @0xmiroslav - looks like this job was auto-doubled, when we stopped this process. Dropo back down to $1000, since we're evaluating proposals?

**Comment by jliexpensify** [comment]
Ah yeah, good point - you did mention that there were "hacky" proposals haha. Ok, we'll keep it at 2k for now.

**Comment by OlimpiaZurek** [comment]
### Please re-state the problem that we are trying to solve in this issue.

If a user renames a workspace by adding multiple consecutive spaces and then opens a modal/dropdown to create a new room, the workspace name will only be displayed with one space instead of the original sequence of multiple consecutive spaces.

### What is the root cause of that problem?

The root cause of this issue is due to the use of the `@react-native-picker/picker` lib in the Expensify app. The library works fine on the native side because it uses native code to manage dropdowns on Android and iOS. However, on the web, a dropdown list is implemented using the `<option>` element. By default, the HTML` <option>` element merges consecutive spaces into one space, as per the HTML standard, and cannot be changed.

So there is no way to avoid modifying the text to solve this problem, but instead of adding changes to the Expensify code I recommend adding it to `@react-native-picker/picker`

It is not possible to avoid platform specific code either as `@react-native-picker/picker` already has platform-dependent code.


### What changes do you think we should make in order to solve the problem?

<ins>Solution # 1<ins>


Given that the dropdown implementation is an integral part of @react-native-picker/picker, the solution should be integrated with this library.

To achieve this, we could add an extra helper function (in `js/PickerItem.js`) to `@react-native-picker/picker`. This property would allow the correct label to be displayed in the <option> element

```
/// This function uses a regular expression to replace all spaces in the label string with the &nbsp; entity. The \u00a0 is the Unicode character for the non-breaking space entity.

const preserveSpaces = (label) =>  {
   return label.replace(/ /g, '\u00a0');
}

<PickerItem
   label={() => preserveSpaces(label)}
/>
```

<ins> Solution # 2 </ins>

I researched various libraries that could potentially replace `@react-native-picker/picker`.  Unfortunately, there is currently no reliable library available that is cross-platform and supports react native for web.

Given that `@react-native-picker/picker` seems to work fine for native part of the app, I'm considering enabling a custom dropdown for the web version. This approach would give us more control over the functionality of the dropdown list and allow us to handle whitespace as needed. 

<ins>Solution # 3 </ins>

The third suggestion is to add a boolean prop for Picker called eg  `preserveSpacesInLabel` (web only) which will implement proposition # 1.

```
const preserveSpaces = (label) => {
 return label.replace(/ /g, '\u00a0');
}
<PickerItem
  label={() => preserveSpacesInLabel ? preserveSpaces(label) : label}
/>

// usage in the App

 <Picker preserveSpacesInLabel />
```

**Comment by alex-mechler** [comment]
Thanks for the investigation @OlimpiaZurek. Proposal 1 looks like the best way to continue forward, as it puts the solution at the library level instead of in Expensify/App. We may want to include the prop as you mentioned in proposal 3, as it would be a behavior change for web if people were not expecting it.

**Comment by cubuspl42** [comment]
@alex-mechler @0xmiroslav

> We may want to include the prop as you mentioned in proposal 3, as it would be a behavior change for web if people were not expecting it.

Sure, that's an upstream review thing.

But details aside, how are we proceeding with proposal acceptance?

I filed my proposal 7 days ago, 4 days ago the [status](https://github.com/Expensify/App/issues/16329#issuecomment-1493645641) was "if that proposal is acceptable, we can assign @cubuspl42", 3 days ago @OlimpiaZurek filed her proposal which comes to the same conclusions as I did.

This clearly is the kind of issue where it's mostly about the idea. Time-wise it's the investigation that takes the most, while the solution itself is a few lines of code.

I'll respect any decision, I just want to know how to approach similar situations in the future.

**Comment by alex-mechler** [comment]
> But details aside, how are we proceeding with proposal acceptance?

We talked a bit internally about this. We definitely gave you conflicting info between here and the other issue. Since this issue didn't have `Help Wanted` when you posted your proposal, we are going to leave this issue in the hands of @OlimpiaZurek. Going forward, please keep proposals to issues that still have the `Help Wanted` label on them.

That said, you did do work on investigating this as a result of our conflicting info, so we would like to pay you $500 for your time. @jliexpensify can get you sorted here. Sorry for any confusion, and thanks for your hard work.

**Comment by jliexpensify** [comment]
@cubuspl42 - I tried looking for you on Upworks but couldn't see your profile. Could you share it with me?

**Comment by cubuspl42** [comment]
@alex-mechler 

Thank you for your explanation!

I created [a PR](https://github.com/Expensify/App/pull/17117) to clarify the rules about the "Help wanted" label.

@jliexpensify

Sure!

https://www.upwork.com/freelancers/~012b13cc109598f579

@OlimpiaZurek 

You might consider (in addition to submitting the upstream PR) temporarily placing the solution [here](https://github.com/Expensify/react-native-picker-select/blob/master/src/index.js#L580), with a comment like "Remove this logic once the upstream PR <link> is merged". That might be enough to close this issue. Otherwise we'd have to either fork `react-native-picker/picker` or put the issue on hold on the upstream review.

**Comment by OlimpiaZurek** [comment]
Thanks @alex-mechler . I will start working on it on Tuesday (as Monday is Bank Holiday in Poland). I will notify you when I submit a PR in `react-native-picker/picker`

**Comment by jliexpensify** [comment]
Paid Jakub :)

**Comment by MelvinBot** [comment]
### Bug0 Triage Checklist ([Main S/O](https://stackoverflowteams.com/c/expensify/questions/14418))

- [ ] This "bug" occurs on a supported platform (ensure `Platforms` in OP are ✅)
- [ ] This bug is not a duplicate report (check [E/App issues](https://github.com/Expensify/App/issues) and [#expensify-bugs](https://expensify.slack.com/archives/C049HHMV9SM))
    - If it is, comment with a link to the original report, close the issue and add any novel details to the original issue instead
- [ ] This bug is reproducible using the reproduction steps in the OP. [S/O](https://stackoverflowteams.com/c/expensify/questions/14418#:~:text=The%20bug%20is%20reproducible)
    - If the reproduction steps are clear and you're unable to reproduce the bug, check with the reporter and QA first, then close the issue.
    - If the reproduction steps aren't clear and you determine the correct steps, please update the OP.
- [ ] This issue is filled out as thoroughly and clearly as possible
    - Pay special attention to the title, results, platforms where the bug occurs, and if the bug happens on staging/production.
- [ ] I have reviewed and subscribed to the linked Slack conversation to ensure Slack/Github stay in sync

**Comment by OlimpiaZurek** [comment]
I've added a temporary solution to the upstream PR here: https://github.com/Expensify/react-native-picker-select/pull/9

@alex-mechler Let me know if this is enough to close this issue or should we wait for the PR at @react-native-picker/picker to be merged. Thanks!

**Comment by alex-mechler** [comment]
I've merged the temporary solution PR. Can you spin up a App PR to update the package we are using?

**Comment by OlimpiaZurek** [comment]
@alex-mechler Thank you. I just opened a PR https://github.com/Expensify/App/pull/17456 .

**Comment by MelvinBot** [comment]
`Reviewing` label has been removed, please complete the "BugZero Checklist".

**Comment by MelvinBot** [comment]
The solution for this issue has been :rocket: deployed to production :rocket: in version 1.3.1-3 and is now subject to a 7-day regression period :calendar:. Here is the list of pull requests that resolve this issue:

- https://github.com/Expensify/App/pull/17456


If no regressions arise, payment will be issued on 2023-04-26. :confetti_ball:

After the hold period is over and BZ checklist items are completed, please complete any of the applicable payments for this issue, and check them off once done.
- [ ] External issue reporter
- [ ] Contributor that fixed the issue
- [ ] Contributor+ that helped on the issue and/or PR

As a reminder, here are the bonuses/penalties that should be applied for any External issue:
- Merged PR within 3 business days of assignment - 50% **bonus**
- Merged PR more than 9 business days after assignment - 50% **penalty**

**Comment by sophiepintoraetz** [comment]
Summarising payments here:
@tienifr : $250 for reporting
@OlimpiaZurek / @0xmiroslav : I'll need your help in summarising what's due here, given there was a period of time between contributor assignment and when the PR was merged?)

(also @alex-mechler if you have any thoughts on this one, seeing as I'm taking over from JLi)

Also, want to make sure there's nothing happening with this [PR here](https://github.com/react-native-picker/picker/pull/485) that is still open?

**Comment by alex-mechler** [comment]
> @OlimpiaZurek / @0xmiroslav : I'll need your help in summarising what's due here, given there was a period of time between contributor assignment and when the PR was merged?)

@OlimpiaZurek is from CallStack, so they will be paid through that https://stackoverflowteams.com/c/expensify/questions/15868/15869#15869

> Also, want to make sure there's nothing happening with this https://github.com/react-native-picker/picker/pull/485 that is still open?

That's an upstream PR where we are held on the maintainer of that library reviewing that pr. We have fixed the issue in Expensify's code, so we should not be waiting on that

**Comment by sophiepintoraetz** [comment]
Cool - Do we have a shortlist somewhere of who is with Callstack? I'll still need help with @0xmiroslav's payment though, @alex-mechler 🙏

**Comment by sophiepintoraetz** [comment]
Okay so @0xmiroslav - you're being compensated $2000? What I'm questioning is the bonus timeframe - are we saying this does not apply because a Callstack engineer created a PR? 
Based on `Agency-employee submits PR and C+ is assigned to review.  C+ is compensated for the job amount.`

**Comment by sophiepintoraetz** [comment]
Okay I'm going to leave the payment as is for now, (no bonus but no penalty either) as it's not clear when the first PR was merged vs contributor assigned.

**Comment by sophiepintoraetz** [comment]
Offer sent, @0xmiroslav!
